### ISSUE TITLE
Early (deny) warnings can hide later actual errors

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#52971]
The following should cause an error on the nonexhaustiveness of `match x`

```rust
#![deny(warnings)]

fn main() {
    let x = Some(42);
    let y = match x {};
    let z = y;
}
```

([Playground](http://play.rust-lang.org/?gist=f91918116cf896ce1cb55e330e355cc7&version=stable&mode=debug&edition=2015))

Errors:

```
   Compiling playground v0.0.1 (file:///playground)
error: unreachable statement
 --> src/main.rs:6:5
  |
6 |     let z = y;
  |     ^^^^^^^^^^
  |
note: lint level defined here
 --> src/main.rs:1:9
  |
1 | #![deny(warnings)]
  |         ^^^^^^^^
  = note: #[deny(unreachable_code)] implied by #[deny(warnings)]

error: aborting due to previous error

error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

I think we should not allow warnings to ever cause `abort_if_errors` to abort. Instead they should fail the build at the earliest after item collection is done (so just before trans starts doing its thing)

**Comment by oli-obk** [comment]
dupe of https://github.com/rust-lang/rust/issues/36161
